src/com/facebook/buck/android/AndroidInstrumentationTest.java

```diff
@@ -0,0 +1,261 @@
+import com.facebook.buck.step.Step;
+import com.facebook.buck.step.fs.MakeCleanDirectoryStep;
+import com.facebook.buck.test.TestCaseSummary;
+import com.facebook.buck.test.TestResults;
+import com.facebook.buck.test.TestResultSummary;
+import com.facebook.buck.test.XmlTestResultParser;
+import com.facebook.buck.test.result.type.ResultType;
+import com.facebook.buck.test.selectors.TestSelectorList;
+import com.facebook.buck.util.Console;
+import com.google.common.base.Functions;
+import com.google.common.base.Optional;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+
+import com.android.ddmlib.IDevice;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+@SuppressWarnings("PMD.TestClassWithoutTestCases")
+public class AndroidInstrumentationTest extends AbstractBuildRule
+    implements TestRule {
+
+  @AddToRuleKey
+  private ImmutableSet<BuildRule> sourceUnderTest;
+
+  private final ImmutableSet<Label> labels;
+
+  private final ImmutableSet<String> contacts;
+
+  private final InstallableApk baseApk;
+
+  private final Optional<AndroidInstrumentationApk> instrumentationApk;
+
+  private final Optional<Long> testRuleTimeoutMs;
+
+  protected AndroidInstrumentationTest(
+      BuildRuleParams params,
+      SourcePathResolver resolver,
+      InstallableApk baseApk,
+      Optional<AndroidInstrumentationApk> instrumentationApk,
+      Set<Label> labels,
+      Set<String> contacts,
+      ImmutableSet<BuildRule> sourceUnderTest,
+      Optional<Long> testRuleTimeoutMs) {
+    super(params, resolver);
+    this.baseApk = baseApk;
+    this.instrumentationApk = instrumentationApk;
+    this.sourceUnderTest = sourceUnderTest;
+    this.labels = ImmutableSet.copyOf(labels);
+    this.contacts = ImmutableSet.copyOf(contacts);
+    this.testRuleTimeoutMs = testRuleTimeoutMs;
+  }
+
+  @Override
+  public ImmutableSet<Label> getLabels() {
+    return labels;
+  }
+
+  @Override
+  public ImmutableSet<String> getContacts() {
+    return contacts;
+  }
+
+  @Override
+  public ImmutableSet<BuildRule> getSourceUnderTest() {
+    return sourceUnderTest;
+  }
+
+  @Override
+  public boolean supportsStreamingTests() {
+    return false;
+  }
+
+  @Override
+  public ImmutableList<Step> runTests(
+      BuildContext buildContext,
+      ExecutionContext executionContext,
+      boolean isDryRun,
+      boolean isShufflingTests,
+      TestSelectorList testSelectorList,
+      TestRule.TestReportingCallback testReportingCallback) {
+    ImmutableList.Builder<Step> steps = ImmutableList.builder();
+
+    Path pathToTestOutput = getPathToTestOutputDirectory();
+    steps.add(new MakeCleanDirectoryStep(pathToTestOutput));
+    steps.add(new ApkInstallStep(baseApk));
+    if (instrumentationApk.isPresent()) {
```

Comment:
I feel like it makes more semantic sense to make baseApk optional and instrumentation_apk as required.

The reason is that the instrumentation_apk is the one that defines <instrumentation> in the manifest, and then in the self-instrumenting case it somehow feels more right to still call that an instrumentation_apk.